### PR TITLE
Training workload for karpathy/nanoGPT in Jax

### LINKED ISSUE #229: ☂️ Training workload for karpathy/nanoGPT
Train karpathy/nanoGPT using technique Full-model training

This training workload is necessary to showcase Parity with tt-train, optimizer benchmarking

### PR DESCRIPTION
**Pull request by ugljesak**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-blacksmith/issues/229)

### Problem description
Bringdown Karpathy's NanoGPT in JAX/Flax and train it on the Tenstorrent N150 hardware. 

### What's changed
I've implemented model in JAX, but for early functionality of `tt-mlir` compiler there are lots of boundaries. I have `nn.Embedding` from JAX which are currently not supported (I'm not sure for the latest version) so there is flag which implements basic MatMul which is a lot slower but it's currently working. Also there is problem with softmax which uses scatter and gather which were not supported by `tt-mlir`, but I got informed that has been fixed in the latest version of compiler. I will test it now, but I need to prolong my N300s access. Still fixing and need testing with new features (Training and providing an example was impossible due to my workaround, using onehot tensor or even worse doing everything on CPU).

### Checklist
- [x] Implemented model
- [x] Tested eval/train with base one on PyTorch using CPU
- [x] Moving to N150 TPU hardware and to run it on tt-mlir compiler.
- [x] Testing training and comparing it to original implementation (waiting for access)
- [x] Fixing potential problems if not matching. 

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212832436210893

### CODE REVIEW

**Review by mmilosevicTT — CHANGES_REQUESTED** [review]
Hi @ugljesak,

Thanks for taking initiative to work on this bounty. The work you provided is great, but it needs a bit of steering to align to the style of experiments that we do in TT-Blacksmith using JAX.

I left some comments that need to be addressed. To help you prioritise, I will provide a summary of larger steps that need to be taken:
- Reworking loading of dataset to align to hugging face loading as in other experiments.
- Reworking experiments to actually use the dataset loaded.
- Reworking environment setup to align to other experiments.
- Updating comparison metrics between Torch and JAX implementations. We need to make sure implementation follows the same pattern on CPU, but comparing Torch/JAX on TT and CPU is redundant for Model correctness.
- Testing with latest version of TT-XLA to see if mentioned ops are still failing.
- Reworking training loop to have at least similar structure to lora/dora experiments for llama. This is important so that we can actually maintain the code later.
- Testing if the loss is actually decreasing for your implementation.

Once you do this, we will be ready to actually run experiments on both CPU and TT, and then compare the results. Since the size of the PR is coming to really large number of lines, I would also suggest splitting addition of model implementation and the experiment itself into two separate PRs.

If you have any more questions, feel free to ping us on this PR by either a message or requesting another review.

**Comment by ugljesak** [comment]
Thank you for the guidance and advices. I got access to N300s back, so I'm back working, will ask you for anything in detail if I get stuck.

**Comment by jberkowitzTT** [comment]
Hi @ugljesak ! Are you blocked on anything are you able to continue working on this PR?

**Comment by jberkowitzTT** [comment]
No problem, thank you for the update!

File: blacksmith/experiments/jax/BOUNTIES/nanogpt/model_jax.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    vocab_size: int = 50304 #5 originally 50257 for gpt2, but padded to the minimal higher multiple of 64 for effeciency
+    num_layers: int = 12
+    num_heads: int = 12
+    num_embeds: int = 768
+    dropout_rate: float = 0.0
+    use_bias: bool = True
+    use_matmul_embed: bool = False
+    dtype: Optional[str] = None
+
+class MatMulEmbed(nn.Module):
```
[the remark this answers is not in the document]
Comment (ugljesak) [reply blacksmith/experiments/jax/BOUNTIES/nanogpt/model_jax.py]: I've pulled newer version of repo, that's working now.

**Comment by ugljesak** [comment]
I'm going to file a bug report on git now, and I've talked to @ndrakulicTT about it, and when it gets resolved, or I make some workaround, it should be a job for a few days to wrap everything up. Sorry for late response.

**Comment by ugljesak** [comment]
Sorry, this week was busy with University, and few more days, then I will try again to run model, and if same old problem keeps happening, I will contact someone to get next steps.

File: blacksmith/experiments/jax/BOUNTIES/nanogpt/dataset.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import os
+import requests
+import numpy as np
+# NO tiktoken. We want characters, not BPE.
+
+input_file_path = os.path.join(os.path.dirname(__file__), 'data/input.txt')
+if not os.path.exists(input_file_path):
+    data_url = 'https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt'
```
[the remark this answers is not in the document]
Comment (ugljesak) [reply blacksmith/experiments/jax/BOUNTIES/nanogpt/dataset.py]: Quick update this was fixed few weeks ago, from now.

File: blacksmith/experiments/jax/BOUNTIES/nanogpt/output/log.csv [hunk 3]
```diff
@@ -0,0 +1,15 @@
+step,train_loss,val_loss,time_sec
```
[the remark this answers is not in the document]
Comment (ugljesak) [reply blacksmith/experiments/jax/BOUNTIES/nanogpt/output/log.csv]: It was slow then and there were some error due to tt-ir compiler having unstable number while processing mathematical operations like pow, now I'm trying to reproduce real results on tiny-shakespeare dataset, where we can se obvious convergence.

File: blacksmith/experiments/jax/BOUNTIES/nanogpt/dataset.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import numpy as np
+from datasets import load_dataset
+
+class ShakespeareDataset:
+    def __init__(self):
+
+        # 1. Download dataset.
+        print("Loading tiny_shakespeare from Hugging Face...")
+        ds = load_dataset("text", data_files={"train": "https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt"})
```
Comment (mmilosevicTT) [inline blacksmith/experiments/jax/BOUNTIES/nanogpt/dataset.py]: why not use 
```
load_dataset("karpathy/tiny_shakespeare", split="train")
```

**Review by mmilosevicTT — COMMENTED** [review]
Hey @ugljesak,

Thanks for taking time to work on this.

I left some comments on the structure of the code. In general this looks like a good quality training example, we just need to polish it a bit and finish. As this is a complex PR, we would need to get couple of eyes on it, but please address these comments in the meantime.

File: blacksmith/experiments/jax/BOUNTIES/nanogpt/train_sp.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+with open(log_file_path, mode='w', newline='') as f:
+    writer = csv.writer(f)
+    writer.writerow(['step', 'train_loss', 'val_loss', 'time_sec'])
+
+    for iter in range(max_iters):
+        
+        # Fetch Batch
+        xb, yb = get_batch('train', train_data, val_data, config.block_size, batch_size, tt_device) 
+        
+        # Compute Gradients (Iter 0 will pause here for ~18s to compile)
```
Comment (mmilosevicTT) [inline blacksmith/experiments/jax/BOUNTIES/nanogpt/train_sp.py]: Timings really depend on the chip used and there might be further improvements to the compiler/hardware, so I would not leave these numbers here.

**Review by mmilosevicTT — APPROVED** [review]
Hey @ugljesak,

LGTM

Please address one last comment and make pre-commit checks pass. I am fine with keeping both scripts (`train_sp` and `test_nanogpt_jax`), as one was used to generate the output.

**Comment by ugljesak** [comment]
Hello @mmilosevicTT @jberkowitzTT 
I have just one more question about mails I've been getting about **nighty uplift workflow runs**:
[ugljesak/tt-blacksmith] Run failed: Nighty Uplift - main (ffd82d4)
I'm still getting them every morning, which is normal I suppose, but now when my branch is merged, is there a way to disable them, or do I need to delete my current fork, in order to achieve that.
If you can explain these tests little better and give me possible solution I would be grateful.
Thanks in advance!
